### PR TITLE
Add undo/redo functionality, update docstrings

### PR DESCRIPTION
**Pull request by blxsse**
Added docstrings to backend functions and added an undo/redo button to the UI.

### CODE REVIEW

File: js/game.js [hunk 1]
```diff
@@ -121,7 +136,7 @@ function game_do_gravity(state) {
 }
 
 function game_try_move_block(state, y, x, dy, dx, slide_rule) {
-    // (State, int, int, int, int, SLIDE_) -> bool
+    // (State, int, int, int, int, SLIDE_, bool) -> bool
```
Comment (diomidov) [inline js/game.js:156]: `slide_rule` is not a `bool`, it's one of `SLIDE_NONE`, `SLIDE_MOVED`, or `SLIDE_ALL`

Comment (diomidova) [inline js/game.js:156]: Where is the 7th bool argument? This function only has 6...

**Review by edemaine — COMMENTED** [review]
I think what we really want is unlimited undo/redo (as in most puzzle games), not just a single operation. Any chance you could extend to an undo stack?

**Review by edemaine — COMMENTED** [review]
Also @diomidov points out:

> If I push a block, and then try to undo it, what causes the block to move back?

You'll need to store the entire game state in the stack, or at least which blocks moved.

**Comment by blxsse** [comment]
Makes sense, thanks so much! I'll work on adding that.

**Pull request closed without merging**